File: server/src/models/category/category.model.ts
import mongoose from "mongoose";
import { ICategoryDocument, CategoryModelType, TypeOfCategory } from "../../lib/types/index";
import { createCategoryPreSave, deleteCategory, deleteCategoryBulk, rebuildTree, restoreCategory } from "./category.utils";

const CategorySchema = new mongoose.Schema<ICategoryDocument, CategoryModelType>(
    {
        name: { type: String, required: true },
        slug: { type: String, required: true },
        subtitle: { type: String },
        thumbnail: { type: String },
        banner: { type: String },
        metaTitle: { type: String },
        metaKeywords: { type: String },
        metaDescription: { type: String },
        markup: { type: String },
        type: { type: String, enum: Object.values(TypeOfCategory), default: TypeOfCategory.CATEGORY, },
        parent: { type: mongoose.Schema.Types.ObjectId, ref: "Category", default: null },
        ancestors: [{ type: mongoose.Schema.Types.ObjectId, ref: "Category" }],
        path: [{ type: String }],
        fullSlug: { type: String },
        level: { type: Number, default: 0 },
        order: { type: Number, default: 0 },
        isActive: { type: Boolean, default: true },
        isDeleted: { type: Boolean, default: false },
        deletedAt: { type: Date, default: null },
    },
    { timestamps: true }
);

CategorySchema.index({ parent: 1, order: 1 });
CategorySchema.index({ type: 1 });
CategorySchema.index({ slug: 1 }, { unique: true, collation: { locale: "en", strength: 2 }, partialFilterExpression: { isDeleted: false } });
CategorySchema.index({ name: "text", metaTitle: "text", metaKeywords: "text" });
CategorySchema.index({ parent: 1, slug: 1 }, { unique: true });
CategorySchema.index({ ancestors: 1 });
CategorySchema.index({ path: 1 });
CategorySchema.index({ fullSlug: 1 }, { unique: true, sparse: true });
CategorySchema.index({ level: 1 });
CategorySchema.index({ isDeleted: 1, isActive: 1, type: 1 });

CategorySchema.pre("save", createCategoryPreSave);

CategorySchema.static('deleteCategory', deleteCategory);
CategorySchema.static('restoreCategory', restoreCategory)

CategorySchema.static('rebuildTree', rebuildTree);

CategorySchema.static('deleteCategoryBulk', deleteCategoryBulk)

export const Category = mongoose.model<ICategoryDocument, CategoryModelType>("Category", CategorySchema);

File: user_app/lib/features/models/category/category_model.dart
import 'package:Quickzon/core/utils/enums/enums.dart';
import 'package:Quickzon/features/models/models.dart';
import 'package:json_annotation/json_annotation.dart';

part 'category_model.g.dart';

@JsonSerializable(explicitToJson: true)
class CategoryType extends CategoryBase {
  @JsonKey(name: '_id')
  final String id;
  final List<CategoryType>? children;
  final DateTime createdAt;
  final DateTime updatedAt;

  @JsonKey(name: '__v')
  final int? v;

  CategoryType({
    required this.id,
    this.children,
    required this.createdAt,
    required this.updatedAt,
    this.v,
    required super.name,
    required super.slug,
    super.subtitle,
    super.thumbnail,
    super.banner,
    super.metaTitle,
    super.metaKeywords,
    super.metaDescription,
    super.markup,
    required super.type,
    super.ancestors,
    super.path,
    super.fullSlug,
    super.level,
    required super.order,
    required super.isActive,
    required super.isDeleted,
    super.deletedAt,
  });

  factory CategoryType.fromJson(Map<String, dynamic> json) =>
      _$CategoryTypeFromJson(json);

  Map<String, dynamic> toJson() => _$CategoryTypeToJson(this);
}
